File: javascript/git_publish.js
#!/usr/bin/env node
// git_publish.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/git-publish-agent/scripts/git_publish.py
// auch in: OpenClaw@gateway2:skills/git-publish-agent/scripts/git_publish.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

/** Git Publish Agent - Automatisierte Skill-Veröffentlichung */

const { execSync, spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

const SKILLS_DIR = path.join(require('os').homedir(), '.openclaw', 'workspace', 'skills');

function runCommand(command, cwd = null) {
  try {
    const result = execSync(command, { 
      cwd: cwd || process.cwd(),
      encoding: 'utf8'
    });
    return { success: true, stdout: result.trim(), stderr: '' };
  } catch (error) {
    return { 
      success: false, 
      stdout: error.stdout ? error.stdout.toString().trim() : '', 
      stderr: error.stderr ? error.stderr.toString().trim() : error.message 
    };
  }
}

function gitCommit(skillPath, message = null) {
  /** Commit skill changes. */
  if (!message) {
    const now = new Date().toISOString();
    message = `[skill] Auto-update ${path.basename(skillPath)} - ${now}`;
  }
  
  const parentDir = path.dirname(SKILLS_DIR);
  runCommand(`git add "${skillPath}"`, parentDir);
  
  const result = runCommand(`git commit -m "${message}"`, parentDir);
  return result.success;
}

function clawhubPublish(skillName) {
  /** Publish to ClawHub. */
  const skillPath = path.join(SKILLS_DIR, skillName);
  const command = `clawhub publish "${skillPath}" --slug ${skillName} --version 1.0.0`;
  
  const result = runCommand(command);
  return [result.success, result.stdout];
}

function batchPublish() {
  /** Publish all changed skills with rate limiting. */
  // Check git status
  const result = runCommand(`git status --short "${SKILLS_DIR}"`);
  
  const changed = [];
  const lines = result.stdout.split('\n');
  
  for (const line of lines) {
    if (line.trim() && line.includes('skills/')) {
      const parts = line.split('skills/');
      if (parts.length > 1) {
        const skill = parts[1].split('/')[0];
        if (!changed.includes(skill)) {
          changed.push(skill);
        }
      }
    }
  }
  
  console.log(`Changed skills: ${JSON.stringify(changed)}`);
  
  // Publish with delay
  const maxBatch = Math.min(5, changed.length);
  for (let i = 0; i < maxBatch; i++) {
    const skill = changed[i];
    if (i > 0) {
      console.log('Waiting 15min for rate limit...');
      // In real: await new Promise(resolve => setTimeout(resolve, 900000));
    }
    
    console.log(`Publishing ${skill}...`);
    const commitOk = gitCommit(path.join(SKILLS_DIR, skill));
    if (commitOk) {
      const [pubOk, output] = clawhubPublish(skill);
      console.log(`  ${pubOk ? '✓' : '✗'} ${output}`);
    }
  }
}

function main() {
  const args = process.argv.slice(2);
  let skill = null;
  let all = false;
  let noPublish = false;
  let message = null;
  
  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--skill':
        skill = args[++i];
        break;
      case '--all':
        all = true;
        break;
      case '--no-publish':
        noPublish = true;
        break;
      case '--message':
        message = args[++i];
        break;
    }
  }
  
  if (skill) {
    const skillPath = path.join(SKILLS_DIR, skill);
    if (noPublish) {
      gitCommit(skillPath, message);
    } else {
      gitCommit(skillPath, message);
      clawhubPublish(skill);
    }
  } else if (all) {
    batchPublish();
  } else {
    console.log('Use --skill <name> or --all');
  }
}

if (require.main === module) {
  main();
}

File: perl5/git_publish.pl
#!/usr/bin/perl
# git_publish.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:skills/git-publish-agent/scripts/git_publish.py
# auch in: OpenClaw@gateway2:skills/git-publish-agent/scripts/git_publish.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use Getopt::Long;
use File::Spec;
use Cwd qw(abs_path);
use POSIX qw(strftime);

# Git Publish Agent - Automatisierte Skill-Veröffentlichung

my $SKILLS_DIR = File::Spec->catdir($ENV{HOME}, '.openclaw', 'workspace', 'skills');

sub git_commit {
    my ($skill_path, $message) = @_;
    
    # Default message if none provided
    if (!$message) {
        my $timestamp = strftime("%Y-%m-%dT%H:%M:%S", localtime);
        $message = "[skill] Auto-update " . basename($skill_path) . " - $timestamp";
    }
    
    # Add files to git
    system('git', 'add', $skill_path) == 0 or die "Failed to add files\n";
    
    # Commit changes
    open(my $fh, '-|', 'git', 'commit', '-m', $message) or die "Cannot run git commit: $!";
    close $fh;
    return $? == 0;
}

sub clawhub_publish {
    my ($skill_name) = @_;
    my $skill_path = File::Spec->catdir($SKILLS_DIR, $skill_name);
    
    open(my $fh, '-|', 
        'clawhub', 'publish', $skill_path,
        '--slug', $skill_name,
        '--version', '1.0.0'
    ) or die "Cannot run clawhub publish: $!";
    
    my $output = do { local $/; <$fh> };
    close $fh;
    my $success = $? == 0;
    
    return ($success, $output);
}

sub batch_publish {
    # Check git status
    open(my $fh, '-|', 'git', 'status', '--short', $SKILLS_DIR) or die "Cannot run git status: $!";
    my @lines = <$fh>;
    close $fh;
    
    my %changed;
    foreach my $line (@lines) {
        chomp $line;
        next unless $line =~ /\S/;
        if ($line =~ m{skills/([^/]+)}) {
            $changed{$1} = 1;
        }
    }
    
    my @changed_skills = keys %changed;
    print "Changed skills: [" . join(", ", @changed_skills) . "]\n";
    
    # Publish with delay (simulated)
    my $count = 0;
    for my $skill (@changed_skills[0..4]) { # Max 5 per batch
        last if $count >= 5;
        
        if ($count > 0) {
            print "Waiting 15min for rate limit...\n";
            # In real: sleep(900);
        }
        
        print "Publishing $skill...\n";
        my $commit_ok = git_commit(File::Spec->catdir($SKILLS_DIR, $skill));
        if ($commit_ok) {
            my ($pub_ok, $output) = clawhub_publish($skill);
            my $symbol = $pub_ok ? '✓' : '✗';
            print "  $symbol $output\n";
        }
        $count++;
    }
}

sub basename {
    my ($path) = @_;
    my @parts = split '/', $path;
    return pop @parts;
}

sub main {
    my $skill;
    my $all = 0;
    my $no_publish = 0;
    my $message;
    
    GetOptions(
        "skill=s"     => \$skill,
        "all"         => \$all,
        "no-publish" => \$no_publish,
        "message=s"   => \$message
    ) or die "Error in command line arguments\n";
    
    if ($skill) {
        my $skill_path = File::Spec->catdir($SKILLS_DIR, $skill);
        if ($no_publish) {
            git_commit($skill_path, $message);
        } else {
            git_commit($skill_path, $message);
            clawhub_publish($skill);
        }
    } elsif ($all) {
        batch_publish();
    } else {
        print "Use --skill <name> or --all\n";
    }
}

main() unless caller;
